### PR TITLE
WIP: Persistence layer migrations for MFA support #419

### LINKED ISSUE #419: Implement Changes to database schema and persistence layer to support TOTP MFA
This task is about changes to the database schema.

Each account can have either 0 or 1 active authenticator app registered. It should be possible to associate a  label 
with the authenticator app (e.g., "Andrea's Android phone"). 

Metadata like creationTime and lastUpdateTime is also stored.

### PR DESCRIPTION
**Pull request by sam-glendenning**
This PR contains the necessary database migrations and Java model changes to support the inclusion of multi-factor authentication into the IAM. Broadly speaking, the changes encompass:

- new tables for authenticator app secrets and their respective recovery codes (`iam_totp_mfa` and `iam_totp_recovery_code`)
- an additional `iam_authority` called `ROLE_PRE_AUTHENTICATED` which is used by Spring to grant a user access to webpages where they can continue the authentication process after signing in with their basic login credentials but does not grant them full access to the IAM. This was necessary to simulate the "two-step" login process associated with multi-factor authentication.

This PR is related to https://github.com/indigo-iam/iam/pull/441 which focuses on the higher-level implementation of MFA in the IAM.

Closes #419

### CODE REVIEW

**Comment by sam-glendenning** [comment]
At this point, I believe the only errors in this branch are the ones inherited from `iam-spring-update-oct-2021`, as this branch has had its commits merged across.

File: iam-persistence/src/main/resources/db/migration/mysql/V71___add_pre_authenticed_authority.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+INSERT INTO iam_authority(AUTH) VALUES
```
Comment (enricovianello) [inline iam-persistence/src/main/resources/db/migration/mysql/V71___add_pre_authenticed_authority.sql:1]: rename file to pre_authenticated_authority

File: iam-persistence/src/main/resources/db/migration/mysql/V70___totp_mfa.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
+-- TOTP MFA secrets
+
+CREATE TABLE iam_totp_mfa (ID BIGINT AUTO_INCREMENT NOT NULL, active TINYINT(1) default 0 NOT NULL, secret VARCHAR(256) NOT NULL, creation_time DATETIME NOT NULL, last_update_time DATETIME NOT NULL, ACCOUNT_ID BIGINT, PRIMARY KEY (ID));
```
Comment (enricovianello) [inline iam-persistence/src/main/resources/db/migration/mysql/V70___totp_mfa.sql:3]: use 255

File: iam-persistence/src/main/resources/db/migration/h2/V70___totp_mfa.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
+-- TOTP MFA secrets
+
+CREATE TABLE iam_totp_mfa (ID BIGINT IDENTITY NOT NULL, active BOOLEAN NOT NULL, secret VARCHAR(256) NOT NULL, creation_time TIMESTAMP NOT NULL, last_update_time TIMESTAMP NOT NULL, ACCOUNT_ID BIGINT, PRIMARY KEY (ID));
+
+CREATE TABLE iam_totp_recovery_code (ID BIGINT IDENTITY NOT NULL, code VARCHAR(255) NOT NULL, totp_mfa_id BIGINT NOT NULL, PRIMARY KEY (ID));
+
+ALTER TABLE iam_account ADD COLUMN totp_mfa_id BIGINT;
```
Comment (enricovianello) [inline iam-persistence/src/main/resources/db/migration/h2/V70___totp_mfa.sql:7]: avoid adding an extra column to iam_account

**Comment by sam-glendenning** [comment]
Currently working on removing changes to iam_account table and creating a repository for iam_totp_mfa. I'll need to test this and make sure it works properly as well. Was held back by SQL problems today so will finish this tomorrow.

**Comment by sam-glendenning** [comment]
@enricovianello I think this is ready for review again with the suggested changes added. The only failing tests are the ones inherited from Andrea's spring boot upgrade branch. 

The new repository for totp secrets contains only one method, for fetching a totp secret by the IAM account ID it relates to. I couldn't think of any further methods that it needed but let me know if you can think of anything.

**Comment by sam-glendenning** [comment]
This branch has got itself confused following the merge of v1.8.0's commits. There are now dozens more files that are registered as changed but are actually merge conflicts with v1.8.0. This likely happened because this branch was originally based off of iam-spring-update-oct-2021 and v.1.8.0 was also based off of iam-spring-update-oct-2021 but has since changed. This PR assumes that the changes made on v1.8.0 will be overwritten by the "changes" made on here.

Rather than fix this entire thing, I thought it would be easier to close this PR and make a new one for the relatively small number of changes this PR is actually trying to make. You can find this new PR here - https://github.com/indigo-iam/iam/pull/460

**Pull request closed without merging**